docs/plugins/new-backend-system.md

```diff
@@ -0,0 +1,305 @@
+
+On startup the backend will make sure that the services are initialized before being passed to the plugin/module that depend on them.
+ServiceRefs contain a scope which is used to determine if the serviceFactory creating the service will create a new instance scoped per plugin/module or if it will be shared. `plugin` scoped services will be created once per plugin/module and `root` scoped services will be created once per backend instance.
+
+#### Defining a Service
+
+```ts
+import {
+  createServiceFactory,
+  pluginMetadataServiceRef,
+  loggerServiceRef,
+} from '@backstage/backend-plugin-api';
+import { ExampleImpl } from './ExampleImpl';
+
+export interface ExampleApi {
+  doSomething(): Promise<void>;
+}
+
+export const exampleServiceRef = createServiceRef<ExampleApi>({
+  id: 'example',
+  scope: 'plugin', // can be 'root' or 'plugin'
+
+  // The defaultFactory is optional to implement but it will be used if no other factory is provided to the backend.
+  // This is allows for the backend to provide a default implementation of the service without having to wire it beforehand.
+  defaultFactory: async service =>
+    createServiceFactory({
+      service,
+      deps: {
+        logger: loggerServiceRef,
+        plugin: pluginMetadataServiceRef,
+      },
+      // Logger is available directly in the factory as it's a root scoped service and will be created once per backend instance.
+      async factory({ logger }) {
+        // plugin is available as it's a plugin scoped service and will be created once per plugin.
+        return async ({ plugin }) => {
+          // This block will be executed once for every plugin that depends on this service
+          logger.info('Initializing example service plugin instance');
+          return new ExampleImpl({ logger });
+        };
+      },
+    }),
+});
+```
+
+### Overriding Services
+
+In this example replace the default root logger service implementation with a custom one that streams logs to GCP. The `rootLoggerServiceRef` has a `'root'` scope, meaning there are no plugin-specific instances of this service.
+
+```ts
+import {
+  createServiceFactory,
+  rootLoggerServiceRef,
+  LoggerService,
+} from '@backstage/backend-plugin-api';
+
+// This custom implementation would typically live separately from
+// the backend setup code, either nearby such as in
+//   packages/backend/src/services/logger/GoogleCloudLogger.ts
+// Or you can let it live in its own library package.
+class GoogleCloudLogger implements LoggerService {
+  static factory = createServiceFactory({
+    service: rootLoggerServiceRef,
+    deps: {},
+    async factory() {
+      return new GoogleCloudLogger();
+    },
+  });
+  // custom implementation here ...
+}
+
+// packages/backend/src/index.ts
+const backend = createBackend({
+  services: [
+    // supplies additional or replacement services to the backend
+    GoogleCloudLogger.factory(),
```

Comment:
What happens if multiple overrides to the same service is accidentally provided here? What is the precedence, does the first or last one found win? Never thought about this until now but I'm assuming it's consistent with the frontend api factory list behavior?

Reply:
I believe the last one wins at the moment. We probably want to keep it that way tbh, to make it a bit easier to have internal defaults and overrides too, thinking `services: [...internalServices, extraInternalService]`. It's a bit of a tradeoff with this pattern tbh, in that it's not always entirely clear what service a factory is for. Could be that naming conventions and patterns will help out here. We've discussed having something like `coreServices.logger()` for example.